Add Pentair ScreenLogic integration

dieselrabbit opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Add `screenlogic` integration for Pentair Intellitouch and EasyTouch pool controllers via a Screenlogic gateway.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
N/A - Config Flow only.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/17010

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dieselrabbit committed:
Initial integration scaffold

dieselrabbit committed:
Initial integration scaffold

dieselrabbit committed:
Initial integration scaffold

dieselrabbit committed:
Initial integration scaffold

dieselrabbit committed:
Starting updates to scaffold

dieselrabbit committed:
Starting updates to scaffold

dieselrabbit committed:
Starting updates to scaffold

dieselrabbit committed:
Starting updates to scaffold

dieselrabbit committed:
First working ScreenLogic integration

dieselrabbit committed:
First working ScreenLogic integration

dieselrabbit committed:
Cleanup

Simplify creation of pump breakout sensors.
Prefix all entities with 'ScreenLogic'

dieselrabbit committed:
Support configuration of multiple gateways

Update config flow to support selection of a gateway to configure, or manual entry.
Move scan interval to options.

dieselrabbit committed:
Move requirements to pypi release of screenlogicpy

Slight change to how discovered gateways are stored/listed.
Adjust wording of some gateway selection UI elements.
Cleanup of unused imports.

bdraco committed:
Small cleanups for Home Assistant standards

- dhcp discovery
- Meet config entry unique id requirements
- Fix switch and water heater doing I/O in the event loop
- Store data by entry_id instead of unique_id
- Only use device class if validated
- DRY
- Make unique id meet https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements
- Remove unless __init__ delegation that pylint complained about

dieselrabbit committed:
Bump screenlogicpy to 0.1.1

Update integration name to Pentair ScreenLogic.
Remove local 'configured_instances()', replace with built-in '_async_current_ids()'

dieselrabbit committed:
Remove api consts from local consts.py.
Restore .devcontainer to upstream/dev.
Add a little ectra debug to gateway_select failure.

dieselrabbit committed:
Update requirements

dieselrabbit commented:
Shoutout to @bdraco for some assists and maybe a code review?

bdraco commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,122 @@
+HEAD_MODE_NAMES = HEAT_MODE._names  # pylint: disable=protected-access
+
+MODE_NAME_TO_MODE_NUM = {
+    HEAD_MODE_NAMES[num]: num for num in range(len(HEAD_MODE_NAMES))
```
```suggestion
    HEAT_MODE_NAMES[num]: num for num in range(len(HEAT_MODE_NAMES))
```

bdraco commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,122 @@
+
+SUPPORTED_FEATURES = SUPPORT_TARGET_TEMPERATURE | SUPPORT_OPERATION_MODE
+
+HEAD_MODE_NAMES = HEAT_MODE._names  # pylint: disable=protected-access
```
```suggestion
HEAT_MODE_NAMES = HEAT_MODE._names  # pylint: disable=protected-access
```

bdraco replied:
Can we avoid the `disable=protected-access` with a change to the pypi package?

bdraco commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,122 @@
+    @property
+    def operation_list(self):
+        """All available operations."""
+        return HEAD_MODE_NAMES
```
```suggestion
        return HEAT_MODE_NAMES
```

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+"""Config flow for ScreenLogic."""
+import logging
```
This one is going to fail the CI since we require 100% coverage for config flows.

bdraco replied:
`python3 -m script.scaffold config_flow` can help generate the base tests

dieselrabbit replied:
Comit [b52ded3](https://github.com/home-assistant/core/pull/47933/commits/b52ded3c135330202d72185ac6f4f4c556dd68bb) should take care of this.

dieselrabbit committed:
Update homeassistant/components/screenlogic/water_heater.py

HEAD_MODE_NAMES to HEAT_MODE_NAMES

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/water_heater.py

HEAD_MODE_NAMES to HEAT_MODE_NAMES

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/water_heater.py

HEAD_MODE_NAMES to HEAT_MODE_NAMES

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,205 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Screenlogic from a config entry."""
+    _LOGGER.debug("Async Setup Entry")
+    _LOGGER.debug(entry.data)
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,205 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    _LOGGER.info("Async Unload Entry")
```
```suggestion
```

dieselrabbit committed:
Update homeassistant/components/screenlogic/__init__.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/screenlogic/binary_sensor.py:
```diff
@@ -0,0 +1,59 @@
+    data = hass.data[DOMAIN][config_entry.entry_id]
+
+    for binary_sensor in data["devices"]["binary_sensor"]:
+        _LOGGER.debug(binary_sensor)
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+    """Derive the gateway name from the mac."""
+    return f"Pentair: {short_mac(mac)}"
+
+
+async def _async_has_devices(hass):
+    """Return if there are devices that can be discovered."""
+    return bool(await async_discover_gateways_by_unique_id(hass))
+
+
+_LOGGER.info("Registering discovery flow")
+config_entry_flow.register_discovery_flow(
+    DOMAIN,
+    "Pentair ScreenLogic",
+    _async_has_devices,
+    config_entries.CONN_CLASS_LOCAL_POLL,
+)
```
Not sure this is needed since we have dhcp discovery below
```suggestion
```

dieselrabbit replied:
Will take a look.

bdraco replied:
Need to double check on this though in the morning after some sleep

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+
+    async def async_step_gateway_select(self, user_input=None):
+        """Handle the selection of a discovered ScreenLogic gateway."""
+        _LOGGER.debug("Gateway Select")
+        _LOGGER.debug(
+            [gateway[SL_GATEWAY_NAME] for gateway in self.discovered_gateways.values()]
+        )
```
```suggestion
```

dieselrabbit committed:
Update homeassistant/components/screenlogic/__init__.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+
+    async def async_step_gateway_entry(self, user_input=None):
+        """Handle the manual entry of a ScreenLogic gateway."""
+        _LOGGER.debug("Gateway Entry")
```
```suggestion
```

dieselrabbit committed:
Update homeassistant/components/screenlogic/binary_sensor.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        if user_input is not None:
+            _LOGGER.debug("Options user_input:")
+            _LOGGER.debug(user_input)
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+
+    def __init__(self, config_entry: config_entries.ConfigEntry):
+        """Init the screen logic options flow."""
+        _LOGGER.debug(config_entry)
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    # Generic sensors
+    for sensor in data["devices"]["sensor"]:
+        _LOGGER.debug(sensor)
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/switch.py:
```diff
@@ -0,0 +1,63 @@
+    data = hass.data[DOMAIN][config_entry.entry_id]
+
+    for switch in data["devices"]["switch"]:
+        _LOGGER.info(switch)
```
```suggestion
```

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,122 @@
+    data = hass.data[DOMAIN][config_entry.entry_id]
+
+    for body in data["devices"]["water_heater"]:
+        _LOGGER.info(body)
```
```suggestion
```

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,202 @@
+
+    await coordinator.async_refresh()
+
+    for circuit in coordinator.data["circuits"]:
+        entities["switch"].append(circuit)
+
+    for sensor in coordinator.data["sensors"]:
+        if sensor == "chem_alarm":
+            entities["binary_sensor"].append(sensor)
+        else:
+            if coordinator.data["sensors"][sensor]["value"] != 0:
+                entities["sensor"].append(sensor)
+
+    for pump in coordinator.data["pumps"]:
+        if (
+            coordinator.data["pumps"][pump]["data"] != 0
+            and "currentWatts" in coordinator.data["pumps"][pump]
+        ):
+            entities["pump"].append(pump)
+
+    for body in coordinator.data["bodies"]:
+        entities["water_heater"].append(body)
```
This this data is already accessible and stored in `coordinator.data`, it would be a bit cleaner to have each platform use it directly instead of creating the intermediate data

dieselrabbit replied:
The raw data stored in the coordinator has more fields than are really useful in home assistant. This intermediate data is a filtered down set of just what should be exposed in home assistant. It could be done for each platform in async_setup_entry(), but I felt that is was more readable to have a singular central place to do/see that.

bdraco replied:
Maybe call this `device_data` instead of `entities` as there aren't home assistant entities which is going to trip up future reviews since we have a specific prohibition on storing entities in a shared container.

dieselrabbit replied:
Can definitely rename for clarity.

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/sensor.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/switch.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/water_heater.py

Remove unnecessary debug logging.

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco reviewed (changes requested):
Looks good.  

Normally we ask that only one platform come in with the initial merge, however since I've already been testing the other platforms ahead of this, I think its ok in this case.

The test coverage will be the blocker here.

You can verify coverage of the config flow with

`pytest --cov=homeassistant/components/screenlogic/ --cov-report term-missing -- tests/components/screenlogic/test_*.py`

Let me know if you get stuck.

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Remove discovery flow registration. No longer needed.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Add test_config_flow.py

dieselrabbit committed:
Remove unused import. Black cleanup.
Remove unused import from removing discovery flow registration.
Black formatting cleanup from removing discovery flow registration.

dieselrabbit committed:
Change 'entities' to 'device_data'.

dieselrabbit committed:
Add option flow coverage to test_config_flow.py

dieselrabbit committed:
Disable false pylint unused-import.

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,225 @@
+GATEWAY_SELECT_KEY = "selected_gateway"
+GATEWAY_MANUAL_ENTRY = "manual"
+
+GATEWAY_ENTRY_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_IP_ADDRESS): str,
+        vol.Required(CONF_PORT, default=80): int,
+    }
+)
+
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/config_flow.py:
```diff
@@ -0,0 +1,225 @@
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.device_registry import format_mac
+
+# pylint: disable=unused-import
+from .const import DEFAULT_SCAN_INTERVAL, DOMAIN, MIN_SCAN_INTERVAL
```
```suggestion
from .const import DOMAIN  # pylint: disable=unused-import
from .const import DEFAULT_SCAN_INTERVAL, MIN_SCAN_INTERVAL
```

bdraco commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,121 @@
+        """Return operation."""
+        return HEAT_MODE.GetFriendlyName(self.body["heat_mode"]["value"])
+
+    @property
+    def operation_list(self):
+        """All available operations."""
+        return HEAT_MODE_NAMES
+
```
<img width="380" alt="Screen Shot 2021-03-15 at 8 14 40 AM" src="https://user-images.githubusercontent.com/663432/111201167-96e9df80-8566-11eb-9d09-b60addcfa6b9.png">

Can we filter out modes that the aren't supported (ex `Solar`).  Also, I think `Don't Change` can be removed

dieselrabbit replied:
Define supported. Supported by the API? Supported by that specific configuration of the pool controller? Supported as a standard mode in Home Assistant?
I'll look into removeing `Don't Change`.

bdraco replied:
> Supported by the user's specific configuration of the pool controller 

The four pools I tested it on don't have Solar so I wasn't expect it to show up.

bdraco replied:
I'm wondering if we can tell from `equipment_flags` if it has solar or not.

bdraco replied:
Maybe https://github.com/parnic/node-screenlogic/blob/main/messages/SLControllerConfigMessage.js#L9

bdraco replied:
https://github.com/parnic/node-screenlogic/blob/main/messages/SLControllerConfigMessage.js#L93

Looks like its `equipment_flags & 0x1`

dieselrabbit replied:
I see. 
I had been taking notes here: https://github.com/dieselrabbit/screenlogicpy/blob/master/notes/equip_flags.txt but since I couldn't account for all options (`heater`) from the equipment flags, I was avoiding it all together. 
Would you consider it better to have all options consistently available and the user be responsible for setting the correct one (because the controller will still take the setting) or support the likely majority of use cases, but in a technically broken way, because `heater` will always be available, even if there is no heater?
To me this also has an impact on the future addition of color lights color/shows, as I feel even farther away from understanding how color-capable being available is identified.

bdraco replied:
I think it is fine for the heater to always appear if we can't figure out which flag it is linked to.  It can always be improved later.  Since we can we figure out which flag controls if solar is present, we should hide that when it is not.

Ideally the lighting moves to be implemented on the `light` platform once more functionality is added (colors/etc).  I can provide config dumps on the 4 pools I have access to if that helps.

dieselrabbit replied:
Ok. I'll update the `water_heater` to remove options when we _know_ the equipment doesn't support it.
Config dumps would be great. Even better would be also knowing what major features they have so I can try to align them to equipment_flags.

As for the future `light`, because there appears to be no full RGB setable lighting from Pentair (only specific colors), it seems currently I'd have to implement all the shows **and** colors as effects, unless we're able to specify which colors go into the light color picker?

dieselrabbit replied:
comit [cf9b5e9](https://github.com/home-assistant/core/pull/47933/commits/cf9b5e9f5dd688f094ba071237135a7ab0745f14) should do this.

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Remove unused GATEWAY_ENTRY_SCHEMA.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update homeassistant/components/screenlogic/config_flow.py

Separate out DOMAIN for pylint disable unused-import.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Update config_flow.py test coverage.

dieselrabbit committed:
Update import order.

dieselrabbit committed:
Filter supported heating modes.
Only add what we believe are supported heating modes to the opperation list.

dieselrabbit commented:
All current requests I believe have been addressed. Please let me know if there's anything else.

bdraco commented:
I'll do a run though tomorrow after sleep.

bdraco commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,127 @@
+
+SUPPORTED_FEATURES = SUPPORT_TARGET_TEMPERATURE | SUPPORT_OPERATION_MODE
+
+HEAT_MODE_NAMES = HEAT_MODE._names  # pylint: disable=protected-access
```
Can you make names public in the library to avoid the pylint exception?

dieselrabbit replied:
Yea, though likely I'll want to change it library-side again in the future. Wanting something more elegant there.

dieselrabbit replied:
Should be resolved with commit [9bece4b](https://github.com/home-assistant/core/pull/47933/commits/9bece4b9bb93ae8f30ac4fe6c53f8c10d105f127)

bdraco commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,202 @@
+    async def _async_update_data(self):
+        """Fetch data from the Screenlogic gateway."""
+        try:
+            self.gateway.update()
```
This likely needs to run in the executor as I'm pretty sure it's doing I/o

dieselrabbit replied:
Yup. Resolved with your suggested change, comit [151490f](https://github.com/home-assistant/core/pull/47933/commits/151490fd61078bb086f22f9d644ee38140117d9e).

bdraco commented:
Dumps of the 4 pools I have access to are here https://github.com/bdraco/screenlogic_dumps

bdraco commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,202 @@
+    async def _async_update_data(self):
+        """Fetch data from the Screenlogic gateway."""
+        try:
+            self.gateway.update()
```
```suggestion
            await hass.async_add_executor_job(self.gateway.update)
```

dieselrabbit committed:
Update homeassistant/components/screenlogic/__init__.py

Run gateway update in executor.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Fix hass reference.

dieselrabbit committed:
Bump screenlogicpy to 0.1.2.
Remove protected use of HEAT_MODE._names.

bdraco reviewed (commented):
Looks good.  Retesting now

bdraco commented:
✅ docs
✅ tests
⚠️ review - code
✅ review - tests
⚠️ testing - in progress

bdraco committed:
Tweak to reduce code

bdraco reviewed (commented):
I pushed a small tweak to reduce the amount of platform code.

bdraco reviewed (approved):
LGTM.  

Thanks for this 👍 

If another reviewer has additional comments, please address them in a followup PR

MartinHjelmare commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,128 @@
+            supported_heat_modes.extend([HEAT_MODE.SOLAR, HEAT_MODE.SOLAR_PREFERED])
+        supported_heat_modes.append(HEAT_MODE.HEATER)
+
+        return [HEAT_MODE_NAMES[mode_num] for mode_num in supported_heat_modes]
```
The only allowed modes are the states in the water heater entity integration.

https://developers.home-assistant.io/docs/core/entity/water-heater#states

bdraco replied:
It might be a better fit to use the climate platform with heat only given the number of presets.

dieselrabbit replied:
Looking at the [Architecture repo](https://github.com/home-assistant/architecture/discussions/518), It appears that there is a desire to bring the `water_heater` platform up to par with `climate`. This seems to include adding "presets" functionality. Ideally, I had thought that I should have a device that heats water to be represented by a `water_heater` entity, also because there is no concept of a fan, but it seem like the `water_heater` platform is too new or maybe not yet mature enough, as several aquatic integrations are using the `climate` platform (iaqualink, smarttub). That said, the Pentair pool controllers that the Screenlogic system supports, in theory have the ability to support a cooling mode as well so `climate` may be better suited anyways.

So the questions I have are: 
1. Are we ok with permanently using the `climate` platform to represent a device that doesn't control the air temperature, in this case the _climate_ of a body of water (pool/spa)? I worry about being disruptive to the end user if at some point in the future we have to move back to `water_heater`.
2. If moving over to `climate`, how does this process work with the release of the integration? Will it be held until the temp controls are moved over to `climate`, or will the integration be released as-is and be a breaking change later?
3. Also, when addressing this (and the rest of the changes) in a new PR, should I keep and continue working in the current `dieselrabbit/screenlogic` branch, or delete this branch and create a new one? Still a bit of a git noob.

Thanks for your time.

MartinHjelmare replied:
1. I think it's ok to use climate since we aren't strict with that categorization yet. Water heater would probably fit better but there's no time table for the improvements there.
2. Unless we revert this PR it will go out in the next minor release of HomeAssistant. If we don't want to release it, we could revert and then make a separate target branch based off dev branch and cherry pick this PR to that branch.That way we can continue the work in the separate target branch and make PRs to that.
3. Start in a fresh branch in the next PR based on latest target branch, eg dev branch.

dieselrabbit replied:
1. I'll start moving the `water-heater` over to `climate`.
2. Hopefully I'll get it done before the next release (unknown when that is).
3. Alright. Will delete this and start a new branch.

bdraco replied:
https://calendar.google.com/calendar/u/0/r/week/2021/3/31?eid=NGRvOGNodGE0ZWg2cnM3M2pqMnUxdDZsOWdfMjAyMTAzMzEgcDA3bjk4Z28xMW9uYW1kMDhkMGttcTZqaHNAZw&ctz=America/Los_Angeles&pli=1&sf=true

For it to make the next release it needs to be merged before March 31st

dieselrabbit replied:
Moved over to climate platform. Basic on/off is in the mode, what type of heating is in the preset.

MartinHjelmare commented on homeassistant/components/screenlogic/water_heater.py:
```diff
@@ -0,0 +1,128 @@
+
+    async def async_set_temperature(self, **kwargs) -> None:
+        """Change the setpoint of the heater."""
+        temperature = kwargs.get(ATTR_TEMPERATURE)
```
Temperature is required in the service schema. We don't need to use dict.get.

MartinHjelmare commented on tests/components/screenlogic/test_config_flow.py:
```diff
@@ -0,0 +1,249 @@
+
+    assert result["type"] == "form"
+    assert result["errors"] == {}
+    assert result["step_id"] == "gateway_entry"
```
Please continue each test to either create entry, error or abort result.

MartinHjelmare commented on tests/components/screenlogic/test_config_flow.py:
```diff
@@ -0,0 +1,249 @@
+    entry = MockConfigEntry(domain=DOMAIN, data={}, options=None)
+    entry.add_to_hass(hass)
+
+    result = await hass.config_entries.options.async_init(entry.entry_id)
```
The integration needs to be set up before we start the options flow to make sure that the config flow module is loaded. Use `hass.config_entries.async_setup`.

dieselrabbit replied:
I'm having trouble with this. I have 3 tests of the options flow and I have added
```
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()
```
to each, after adding the mock entry and before calling `async_init()`, as other integrations have done. The first one to run passes fine, but the remaining 2 fail in `async_setup()` with:
```
        if entry.state != ENTRY_STATE_NOT_LOADED:
>           raise OperationNotAllowed
E           homeassistant.config_entries.OperationNotAllowed
```
I'm not sure what I'm doing wrong.

dieselrabbit replied:
I think I resolved this. I was trying to test with a full `MockConfigEntry`, instead of one that only specifies the `domain`.

MartinHjelmare reviewed (commented):
It would have been good to limit the first PR to one platform and have the rest in follow up PRs.

Please address the comments in a new PR. Thanks!

dieselrabbit commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,202 @@
+    @property
+    def device_info(self):
+        """Return device information for the controller."""
+        controller_type = self.config_data["controler_type"]
```
This typo started in the library. Will fix library and integration.

dieselrabbit commented on homeassistant/components/screenlogic/binary_sensor.py:
```diff
@@ -0,0 +1,54 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        device_class = self.sensor.get("hass_type")
```
I wrote the library specifically with Home Assistant in mind, so picking types/device_classes that matched Home Assistant seemed harmless enough, but you are right that serving the CLI with the library and translating in hass would be a better design choice.
This also highlights that somewhere along the way in development of the library, `hass_type` changed to `hass_device_class` and not all previous instances were corrected. I'll resolve both in a future release.

bdraco replied:
Thanks. I recall similar feedback about not having the library know about home assistant device classes was given to the tesla code owner in a previous PR, although we didn't press to hard on switching it as it was existing behavior.  Sorry I missed that on the first round.

dieselrabbit commented on homeassistant/components/screenlogic/switch.py:
```diff
@@ -0,0 +1,63 @@
+        return self.circuits_data[self._data_key]
+
+    @property
+    def circuits_data(self):
```
I believe this was done for consistency with the other platforms. Would you still suggest breaking that consistency to inline it?

MartinHjelmare replied:
It's not a big deal either way I think.

dieselrabbit replied:
As they were only used in that one instance far all the platforms, I ended up inlining it.

dieselrabbit commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,202 @@
+    except ScreenLogicError as ex:
+        _LOGGER.error("Error while connecting to the gateway %s: %s", connect_info, ex)
+        raise ConfigEntryNotReady from ex
+    except AttributeError as ex:
```
@bdraco You added this. Did you ever receive an `AttributeError`? Not sure where that would be raised from.

bdraco replied:
Change the ip address to one that doesn't actually have a screenlogic controller running on it and it should happen.

dieselrabbit replied:
Updated library should error gracefully. Removed `AttributeError` handler.

dieselrabbit commented on homeassistant/components/screenlogic/binary_sensor.py:
```diff
@@ -0,0 +1,54 @@
+
+    for binary_sensor in data["devices"]["binary_sensor"]:
+        entities.append(ScreenLogicBinarySensor(coordinator, binary_sensor))
+    async_add_entities(entities, True)
```
Should I be removing the second argument for all platforms?

bdraco replied:
Yes

bdraco commented on homeassistant/components/screenlogic/__init__.py:
```diff
@@ -0,0 +1,202 @@
+    if mac in discovered_gateways:
+        connect_info = discovered_gateways[mac]
+    else:
+        _LOGGER.warning("Gateway rediscovery failed.")
```
This was fixed on `dev` during a mass cleanup of these.

bdraco commented on homeassistant/components/screenlogic/switch.py:
```diff
@@ -0,0 +1,63 @@
+        if await self.hass.async_add_executor_job(
+            self.gateway.set_circuit, self._data_key, circuit_value
+        ):
+            _LOGGER.info("screenlogic turn %s %s", circuit_value, self._data_key)
```
This was fixed on `dev` during a mass cleanup of these.

dieselrabbit commented:
Everything should be addressed in the new PR.
